### ISSUE TITLE
Add WDM technology-specific constraints

### ISSUE THREAD

**Issue by italobusi** [aguoietf/ietf-ccamp-optical-path-computation#15]
**2021-12-16 Call**

In addition to optical impairments-related constraints (issue #14 ) there can be other WDM technology-specific constraints

Candidates:
- path length or latency
- number of hops

To check if they already exist as a generic constraints

**Comment by ggalimba56** [comment]
In a Flexigrid tunnel it is reasonable to suppose some constraints may be set from the operator,
the most significant are:
- MAX number of hops
- MAX circuit (tunnel) length (latency)
- MAX circuit Noise (or Min OSNR) value.
The Operator should be able to set any combination of the above constraints (giving values) as well as giving the criteria to sort the the available tunnel list (e.g. sorting first on the number of hops or on the circuit length, etc.).
All the above can give an excellent path computation accuracy in terms of optical feasibility and circuit performance.

There are other kind of constraints required by and operator):
- diversity (from an existing circuit)
- exclusion of resources already in use or critical (XRO - e.g. Nodes, Links, SRLG)
- Inclusion of resources or exlicit route (ERO)
- Spectrum assignment.
The above constraints are considered in the path computation process.

**Comment by italobusi** [comment]
Re-opened to track the actions required to support these requirements

**Comment by italobusi** [comment]
**2022-02-24 Call**

> * MAX number of hops
> * MAX circuit (tunnel) length (latency)

Supported by the generic TE tunnel and Path Computation RPC models as metric-bounds

> * MAX circuit Noise (or Min OSNR) value.

Addressed in #14 

> * diversity (from an existing circuit)
> * exclusion of resources already in use or critical (XRO - e.g. Nodes, Links, SRLG)
> * Inclusion of resources or exlicit route (ERO)

Supported by the generic TE tunnel and Path Computation RPC models

> * Spectrum assignment.

Supported by the existing technology-specific augmentations of the TE label containers for the generic TE tunnel and Path Computation RPC models

Additional technology-specific constraints (to be added):
- include/exclude OMS elements

**Comment by italobusi** [comment]
> Additional technology-specific constraints (to be added):
> 
> * include/exclude OMS elements
> 
> This constraint can be configured in either path computation RPC or in the tunnel model (per-path)
> 
> To be added in layer0-types-ext and to be used by all the technology-specific tunnel and path-computation models

We need to augment the following choice with one attribute (oms-element-id):

```
       |  |  |  +-- (type)?
       |  |  |     +--:(numbered-node-hop)
       |  |  |     |  +-- numbered-node-hop
       |  |  |     |     +-- node-id     te-node-id
       |  |  |     |     +-- hop-type?   te-hop-type
       |  |  |     +--:(numbered-link-hop)
       |  |  |     |  +-- numbered-link-hop
       |  |  |     |     +-- link-tp-id    te-tp-id
       |  |  |     |     +-- hop-type?     te-hop-type
       |  |  |     |     +-- direction?    te-link-direction
       |  |  |     +--:(unnumbered-link-hop)
       |  |  |     |  +-- unnumbered-link-hop
       |  |  |     |     +-- link-tp-id    te-tp-id
       |  |  |     |     +-- node-id       te-node-id
       |  |  |     |     +-- hop-type?     te-hop-type
       |  |  |     |     +-- direction?    te-link-direction
       |  |  |     +--:(as-number)
       |  |  |     |  +-- as-number-hop
       |  |  |     |     +-- as-number    inet:as-number
       |  |  |     |     +-- hop-type?    te-hop-type
       |  |  |     +--:(label)
       |  |  |        +-- label-hop
       |  |  |           +-- te-label
       |  |  |              +-- (technology)?
       |  |  |              |  +--:(generic)
       |  |  |              |     +-- generic?
       |  |  |              |             rt-types:generalized-label
       |  |  |              +-- direction?       te-label-direction

``` 

This should be done in each model and not in the layer0-types-ext.

Co-Authored-By: sergio belotti [sergio.belotti@nokia.com](mailto:sergio.belotti@nokia.com)
